subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java

```diff
@@ -61,74 +65,102 @@
-    private final GlobalScopedCacheBuilderFactory cacheBuilderFactory;
     private final ClassLoaderHierarchyHasher classLoaderHierarchyHasher;
     private final CachedClasspathTransformer classpathTransformer;
+    private final ExecutionEngine executionEngine;
+    private final FileCollectionFactory fileCollectionFactory;
+    private final InputFingerprinter inputFingerprinter;
+    private final ImmutableWorkspaceProvider workspaceProvider;
 
     public FileCacheBackedScriptClassCompiler(
-            GlobalScopedCacheBuilderFactory cacheBuilderFactory, ScriptCompilationHandler scriptCompilationHandler,
-            ProgressLoggerFactory progressLoggerFactory, ClassLoaderHierarchyHasher classLoaderHierarchyHasher,
-            CachedClasspathTransformer classpathTransformer) {
-        this.cacheBuilderFactory = cacheBuilderFactory;
+        ScriptCompilationHandler scriptCompilationHandler,
+        ClassLoaderHierarchyHasher classLoaderHierarchyHasher,
+        CachedClasspathTransformer classpathTransformer,
+        ExecutionEngine executionEngine,
+        FileCollectionFactory fileCollectionFactory,
+        InputFingerprinter inputFingerprinter,
+        ImmutableWorkspaceProvider workspaceProvider
+    ) {
         this.scriptCompilationHandler = scriptCompilationHandler;
-        this.progressLoggerFactory = progressLoggerFactory;
         this.classLoaderHierarchyHasher = classLoaderHierarchyHasher;
         this.classpathTransformer = classpathTransformer;
+        this.executionEngine = executionEngine;
+        this.fileCollectionFactory = fileCollectionFactory;
+        this.inputFingerprinter = inputFingerprinter;
+        this.workspaceProvider = workspaceProvider;
     }
 
     @Override
-    public <T extends Script, M> CompiledScript<T, M> compile(final ScriptSource source,
-                                                              final ClassLoaderScope targetScope,
-                                                              final CompileOperation<M> operation,
-                                                              final Class<T> scriptBaseClass,
-                                                              final Action<? super ClassNode> verifier) {
+    public <T extends Script, M> CompiledScript<T, M> compile(
+        final Object target,
+        final ScriptSource source,
+        final ClassLoaderScope targetScope,
+        final CompileOperation<M> operation,
+        final Class<T> scriptBaseClass,
+        final Action<? super ClassNode> verifier
+    ) {
         assert source.getResource().isContentCached();
         if (source.getResource().getHasEmptyContent()) {
-            return emptyCompiledScript(operation);
+            return new EmptyCompiledScript<>(operation);
         }
 
-        ClassLoader classLoader = targetScope.getExportClassLoader();
+        String dslId = operation.getId();
         HashCode sourceHashCode = source.getResource().getContentHash();
```

Comment:
❓ We shouldn't need to calculate our own hash here. Instead, the execution engine could handle it as part of visiting inputs, no? We can opt to keep it like this for now, but in that case I would leave a comment to explain why it is done like this, and how we can utilize the standard input file snapshotting later.

Reply:
Alright. Let's come back to this at some point. Maybe we could call it the same way for Kotlin and Groovy though?